src/sharing/SharingScreen.js

```diff
@@ -0,0 +1,498 @@
+    const to = JSON.stringify(users.map(user => user.user_id));
+
+    try {
+      showToast('Sending Message...');
+
+      if (sharedData.type === 'image' || sharedData.type === 'file') {
+        const url =
+          sharedData.type === 'image' ? sharedData.sharedImageUrl : sharedData.sharedFileUrl;
+        const fileName = url.split('/').pop();
+        const response = await uploadFile(auth, url, fileName);
+        messageToSend += `\n[${fileName}](${response.uri})`;
+      }
+      await sendMessage(auth, { content: messageToSend, type: 'private', to });
+    } catch (err) {
+      showToast('Failed to send message');
+      this.finishShare();
+      return;
+    }
+
+    showToast('Message sent');
+    this.finishShare();
+  };
+
+  render() {
+    const { message, choosingRecipients, sending } = this.state;
+
+    if (choosingRecipients) {
+      return (
+        <Modal>
+          <ChooseRecipientsScreen onComplete={this.handleChooseRecipients} />
+        </Modal>
+      );
+    }
+
+    const { sharedData } = this.props.navigation.state.params;
+    let sharePreview = null;
+    if (sharedData.type === 'image') {
+      sharePreview = (
+        <Image
+          source={{ uri: sharedData.sharedImageUrl }}
+          width={200}
+          height={200}
+          style={styles.imagePreview}
+        />
+      );
+    }
+
+    return (
+      <>
+        <ScrollView style={styles.wrapper} keyboardShouldPersistTaps>
+          <View style={styles.container}>{sharePreview}</View>
+          <View style={styles.usersPreview}>{this.getUsersPreview()}</View>
+          <ZulipButton
+            onPress={this.toggleChoosingRecipients}
+            style={styles.chooseButton}
+            text="Choose recipients"
+          />
+          <Input
+            value={message}
+            placeholder="Message"
+            onChangeText={this.handleMessageChange}
+            multiline
+          />
+        </ScrollView>
+        <View style={styles.actions}>
+          <ZulipButton onPress={this.finishShare} style={styles.button} secondary text="Cancel" />
+          <ZulipButton
+            style={styles.button}
+            onPress={this.handleSend}
+            text="Send"
+            progress={sending}
+            disabled={!this.isSendButtonEnabled()}
+          />
+        </View>
+      </>
+    );
+  }
+}
+
+const ShareToPm = connect(state => ({
+  auth: getAuth(state),
+}))(ShareToPmComponent);
+
+const ShareToStream = connect(state => ({
+  subscriptions: getSubscriptionsById(state),
+  auth: getAuth(state),
+}))(ShareToStreamComponent);
+
+type Props = $ReadOnly<{|
+  navigation: NavigationScreenProp<{ params: {| sharedData: SharedData |} }>,
+  auth: Auth | void,
+  dispatch: Dispatch,
+|}>;
+
+const SharingTopTabNavigator = createMaterialTopTabNavigator(
+  {
+    Stream: {
+      // $FlowFixMe
+      screen: ShareToStream,
+    },
+    'Private Message': {
+      // $FlowFixMe
+      screen: ShareToPm,
+    },
```

Comment:
What's the reason for these fixmes?

In general where there's a `$FlowFixMe`, we always want at least a line or so of comment identifying why it was necessary.

Reply:
Removing the `$FlowFixMe` causes the following error:
```
Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ src/sharing/SharingScreen.js:23:15

Cannot call createMaterialTopTabNavigator with object literal bound to routeConfigs because in property Stream.screen:
 • Either property navigationOptions is missing in React.ComponentType [1] but exists in
   withOptionalNavigationOptions [2].
 • Or React.ComponentType [1] is incompatible with React.StatelessFunctionalComponent [3].

     src/sharing/SharingScreen.js
      18│ |}>;
      19│
      20│ const SharingTopTabNavigator = createMaterialTopTabNavigator(
      21│   {
      22│     Stream: {
      23│       screen: ShareToStream,
      24│     },
      25│     'Private Message': {
      26│       // $FlowFixMe
      27│       screen: ShareToPm,
      28│     },
      29│   },
      30│   {
      31│     tabBarPosition: 'top',
      32│     animationEnabled: true,

     flow-typed/npm/react-navigation_v2.x.x.js
 [2] 313│     withOptionalNavigationOptions<Options>;
        :
 [3] 628│   > = React$StatelessFunctionalComponent<{
     629│     ...Props,
     630│     ...NavigationContainerProps<State, Options>,
     631│   }> &

     src/react-redux.js
 [1]  98│ ): C => ComponentType<$ReadOnly<OwnProps<C, SP>>> {
```

@chrisbobbe do you know what might be causing this? I can't figure out. I've added a generic comment:
```
      // Requires additional props that we dont need. $FlowFixMe.
```

Reply:
I think https://github.com/zulip/zulip-mobile/pull/4114#issuecomment-634255590 is our latest knowledge on this error. Thanks, Greg, for writing that up! 🙂

Reply:
Ah, and that in turn leads me to this existing fixme:
```
$ rg -C2 FixMe.*nav 
src/nav/AppNavigator.js
34-
35-export default createStackNavigator(
36:  // $FlowFixMe react-navigation types :-/ -- see a36814e80
37-  {
38-    account: { screen: AccountPickScreen },
```

Which, when I try removing it, produces exactly the same kind of error as @agrawal-d quoted above.

That more recent writeup has more information than the commit mentioned in that fixme. But that certainly serves as an example of how these explanations don't need to be long :wink: .

A couple of smaller points on what goes in the fixme:
* The keyword `$FlowFixMe` goes at the start of the line, right after the `// `. That helps it stand out when scanning the code by eye. (Same thing goes for TODO, or anything else of that ilk.)
* Thinking about it a bit more, I think the key information I want in the explanation comes down to: why haven't we just already fixed it? So with "Requires additional props that we don't need", the natural next thing is: OK, let's make whatever's requiring those stop requiring them, then. But then the answer to that is: ah but it's over in the react-navigation library (or anyway its types, specified in its libdef). That doesn't mean we *can't* fix it -- it's just code, we can change it -- but it's more work. So identifying a specific third-party place as the location of the problem is a common kind of explanation for these.

Reply:
> Ah, and that in turn leads me to this existing fixme: [...]

One minor point about that example is that the `$FlowFixMe` is less precise than it could be; in this PR and https://github.com/zulip/zulip-mobile/pull/4114/files (in the latter, at Greg's suggestion) we get some nice precision by putting it directly above the `screen` property in all those places. (I just tried that with one case in AppNavigator.js and commented out the other cases, and it was fine, as we'd expect.)

But adding 29 `$FlowFixMe`s might be cumbersome, and we don't use anything other than the `screen` property. In fact, I'm pretty sure (untested though) the API would allow us to do this for all of them there:

```diff
- account: { screen: AccountPickScreen },
+ account: AccountPickScreen,
```